lib/media/segment_prefetch.js

```diff
@@ -0,0 +1,173 @@
+/*! @license
+ * Shaka Player
+ * Copyright 2016 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+goog.require('shaka.net.NetworkingEngine');
+goog.require('shaka.media.InitSegmentReference');
+goog.require('shaka.media.SegmentReference');
+goog.provide('shaka.media.SegmentPrefetch');
+goog.require('shaka.log');
+
+/**
+ * @summary
+ * This class manages segment prefetch operations.
+ * Called by StreamingEngine to prefetch next N segments
+ * ahead of playhead, to reduce the chances of rebuffering.
+ */
+shaka.media.SegmentPrefetch = class {
+  /**
+   * @param {number} prefetchLimit
+   * @param {shaka.extern.Stream} stream
+   * @param {function(
+   *  !(shaka.media.InitSegmentReference|shaka.media.SegmentReference),
+   *  shaka.extern.Stream
+   * ):!shaka.net.NetworkingEngine.PendingRequest} fetchDispatcher
```

Comment:
I would define the `fetchDispatcher` function with a `@typedef`, so it doesn't need to be written out each time like this.
See, for reference, `shaka.net.NetworkingEngine.OnHeadersReceived` inside `lib/net/networking_engine.js`.

Reply:
defined type